Fixes #36971 - GUI to allow cloning of Ansible roles from VCS

Thorben-D opened the pull request:
This feature adds a new modal to /ansible/ansible_roles, which allows the user to clone a git-repo with roles to a specified SmartProxy.

[RedMine-Ticket](https://projects.theforeman.org/issues/36971)

GUI:

![Screenshot_20231208_134233](https://github.com/theforeman/foreman_ansible/assets/99347625/6aa9926b-372e-4730-8148-24d72a8ba036)

**Depends on** https://github.com/theforeman/smart_proxy_ansible/pull/85

nofaralfasi commented:
I'll test it again after the tests are successful because it is not working for me now.

Thorben-D commented:
I fixed the linting issues and retested everything.
Tests are running fine on my machine.

nofaralfasi commented:
I'm getting the following error from the GUI:
```
Performing CloneAnsibleRole (Job ID: 311cb572-ffe0-43c1-be85-3253315eb4d3) from Dynflow(default) enqueued at 2023-12-14T10:15:52Z with arguments: {"vcs_url"=>"https://github.com/theforeman/foreman.git", "name"=>"theforeman.foreman", "ref"=>"", "update"=>false}, #<GlobalID:0x00007fb5c0bda290 @uri=#<URI::GID gid://foreman/SmartProxy/1>>
Error performing CloneAnsibleRole (Job ID: 311cb572-ffe0-43c1-be85-3253315eb4d3) from Dynflow(default) in 65.03ms: Foreman::Exception (ERF42-7661 [Foreman::Exception]: »Git Error. Check log.«):
```

Thorben-D commented:
I added the missing permissions. The pipeline should now complete.

ekohl commented on app/controllers/api/v2/vcs_clone_controller.rb:
```diff
@@ -0,0 +1,80 @@
+      error 500, :desc => N_('Internal server error')
+      def installed_roles
+        smart_proxy = params.require(:smart_proxy)
+        ansible_proxy = SmartProxy.find_by(name: smart_proxy)
```
You should at least require it to have the Ansible feature:
```suggestion
        ansible_proxy = SmartProxy.with_features('Ansible').find_by(name: smart_proxy)
```

But you can also verify a capability:
```suggestion
        ansible_proxy = SmartProxy.find_by(name: smart_proxy)
        raise Exception unless ansible_proxy.has_capability?('Ansible', 'vcs')
```

Thorben-D replied:
Thank you very much for the detailed explanation!
It does indeed make sense to only have one of the two plugins be dependent on git.
Adding a new capability to denote cloning functionality is also a nice idea.

ekohl commented on app/jobs/clone_ansible_role.rb:
```diff
@@ -0,0 +1,14 @@
+  queue_as :default
+
+  def humanized_name
+    'Clone Ansible Role from VCS'
```
Aren't these typically translatable?
```suggestion
    _('Clone Ansible Role from VCS')
```

ekohl commented on app/controllers/api/v2/vcs_clone_controller.rb:
```diff
@@ -0,0 +1,80 @@
+        head :internal_server_error
+      end
+
+      api :GET, '/vcs_clone/get_repo_info', N_('Returns information about the repo')
```
REST encodes the method in the HTTP method, so `get_repo_info` is redundant. See further suggestions in the routes part.

ekohl commented on config/routes.rb:
```diff
@@ -100,6 +100,14 @@
 
+        resources :vcs_clone, :only => [] do
+          collection do
+            get 'get_repo_information', to: 'vcs_clone#repo_information'
+            post 'install_role', to: 'vcs_clone#install_role'
+            get 'get_installed_roles', to: 'vcs_clone#installed_roles'
```
REST encodes methods in the HTTP method, not in the path. So I'd think this is a more natural API:
```suggestion
            get 'repo_information', to: 'vcs_clone#repo_information'
            post 'roles', to: 'vcs_clone#install_role'
            get 'roles', to: 'vcs_clone#installed_roles'
```

You `POST` to `roles` to create a new one and `GET` to `/roles` to get the installed roles.

Then uninstall would be `DELETE /roles/:id` and updating `PUT /roles/:id`.

Thorben-D replied:
Indeed, this makes more sense semantically.
I moved the functionality behind a single endpoint with GET, POST, PUT and DELETE methods.

ekohl commented on foreman_ansible.gemspec:
```diff
@@ -22,4 +22,5 @@ Gem::Specification.new do |s|
   s.add_dependency 'deface', '< 2.0'
   s.add_dependency 'foreman_remote_execution', '>= 9.0', '< 13'
   s.add_dependency 'foreman-tasks', '>= 7.0', '< 10'
+  s.add_dependency 'git', '~> 1.0'
```
I'd prefer to avoid the whole git dependency. See more in the overall review.

Thorben-D replied:
All git functionality has been moved to the smartproxy plugin. I removed the dependency.

ekohl commented on app/controllers/api/v2/vcs_clone_controller.rb:
```diff
@@ -0,0 +1,80 @@
+        param :ref, String, :desc => N_('Branch / Tag / Commit reference'), :required => true
+        param :update, String, :desc => N_('Whether an existing role with the provided name should be updated'), :default => false
+      end
+      param :smart_proxy, Array, N_('Array of SmartProxies the role should get installed to')
```
Why is this an array, but only the first ID gets used?

Thorben-D replied:
This has been indirectly resolved by passing the proxy-id as a url parameter.

ekohl commented on app/controllers/api/v2/vcs_clone_controller.rb:
```diff
@@ -0,0 +1,80 @@
+        payload['repo_info']['update'] = false unless payload['repo_info'].key? 'update'
+
+        # 'smart_proxy' is an array to later allow a role to be installed on multiple SPs
+        ansible_proxy = SmartProxy.find_by(name: payload['smart_proxy'][0])
```
Similar to the above: please verify the feature and capability.

Thorben-D commented:
I forgot to remove the 'git' dependency and foolishly force pushed again after removing it.
That kind of broke the 'Compare' function...
The actual commit that addresses the feedback from @ekohl is ae450111d4c96a6cacfa7c0e6a727b4263f28df0

ekohl commented on app/controllers/api/v2/vcs_clone_controller.rb:
```diff
@@ -0,0 +1,150 @@
+      rescue Foreman::Exception
+        head :internal_server_error
+      end
+
+      private
```
There's already a `private` statement on line 80.
```suggestion
```

Thorben-D replied:
The whole block was useless. Not sure, how I missed that...

ekohl commented on app/lib/proxy_api/ansible.rb:
```diff
@@ -53,5 +53,61 @@ def playbooks(playbooks_names = [])
+      when 400
+        raise Foreman::Exception.new N_('Error requesting repo-info. Check Smartproxy log.')
+      else
+        raise e
```
I think this causes a different trace back and you can simply use it without arguments to raise it again:
```suggestion
        raise
```

ekohl commented on app/lib/proxy_api/ansible.rb:
```diff
@@ -53,5 +53,61 @@ def playbooks(playbooks_names = [])
+
+    def repo_information(vcs_url)
+      parse(get("vcs_clone/repo_information?vcs_url=#{vcs_url}"))
+    rescue *PROXY_ERRORS, RestClient::Exception => e
+      raise e unless e.is_a? RestClient::RequestFailed
```
I'd only catch the specific exception:
```suggestion
    rescue RestClient::RequestFailed => e
```

Thorben-D replied:
I had something different in my mind, but just now realized that your suggestion works fine and is cleaner...
I'll add it after the new changes have been reviewed.

ekohl commented on app/controllers/api/v2/vcs_clone_controller.rb:
```diff
@@ -0,0 +1,150 @@
+        start_vcs_task(payload, :install)
+      end
+
+      api :PUT, '/smart_proxies/:proxy_name/ansible/vcs_clone/roles',
+          N_('Launches a task to update the provided role')
+      formats ['json']
+      param :repo_info, Hash, :desc => N_('Dictionary containing info about the role to be installed') do
```
Semantically I'd expect a `PUT` on a collection to ensure that the provided list of items is the exact set of items present. Or put in other words, `PUT` on `.../roles` with `my_role` would ensure that only `my_role` is present and all other roles are removed.

If the name is immutable (and I think it's safe to assume so) you could make it part of the URL and make it `PUT .../roles/:name` to update things (like VCS URL and ref). Just like you did with `DELETE`.

Thorben-D replied:
I added `role_name` as a parameter to the URL, which does indeed make more sense now!

ekohl reviewed (changes requested):
Thanks for addressing my concerns. I've only looked at the Ruby side of it, not the front end but I think this generally looks good.

I didn't see it in the PR, but for myself: this relies on https://github.com/theforeman/smart_proxy_ansible/pull/85.

Thorben-D commented:
The latest commit splits up the large React-components into smaller, more manageable ones. The Ruby-part stays untouched.

gardar commented:
Excited to see some git integration being added for foreman_ansible, but I have few thoughts.
  - How about syncing collections too? Ansible is moving into the direction of collections and away from standalone roles.
  - Wouldn't it have been better to do the integration with ansible-galaxy? ansible-galaxy can clone both roles and collections from git but also fetch them from a galaxy server.

With that being said, do you have any plans to add further git integration with foreman_ansible? Would love to be able to sync the ansible variables to and from a git repo.

OttaviaB commented:
That is a good idea about collections!
In foreman, collections are still handled in a fundamentally different way than roles though, right?
Roles are imported and assigned to hosts/host groups, while collections must be simply installed once and (at least in my experience) are usually not updated as often as roles.
I guess that the main pain point is with roles for now, but it definetely makes sense to think about collections as well in the future.
 
Using ansible-galaxy instead of git is a great suggestion. Maybe we can do that later?

gardar commented:
> In foreman, collections are still handled in a fundamentally different way than roles though, right?

Yes and no, a collection can be a collection of roles or even a single role. The collection can also include ansible modules, playbooks, etc. All of this is optional so that the collections don't have to contain roles but they also don't have to contain modules, etc.

> Roles are imported and assigned to hosts/host groups, while collections must be simply installed once and (at least in my experience) are usually not updated as often as roles.

By installing a collection that contains roles on a foreman server you get ansible roles that you can import in foreman and can then assign those to hosts/host groups.

It's absolute nonsense that collections are usually not updated as often as roles, especially when collections can include roles.

When roles are packaged in a collection you can't install just the roles unless you manually copy the roles out of the collection which is something that should be avoided at all cost.
I think it's more trouble than it's worth to do this without using ansible-galaxy as ansible-galaxy already handles the roles and collections correctly.

OttaviaB commented:
Thanks for your feedback. While this PR aims to streamline a specific workflow we see being commonly used, we will think about whether we want to address your suggestions in a later PR.

sbernhard commented:
@nofaralfasi Would it be possible for you to have a look at this PR or assign it to someone else?

nofaralfasi commented:
> @nofaralfasi Would it be possible for you to have a look at this PR or assign it to someone else?

Currently, I am unable to test the PR due to some errors I am encountering (see my comment here: https://github.com/theforeman/smart_proxy_ansible/pull/85/files#r1510263657).

nofaralfasi reviewed (commented):
I'm encountering some issues when I navigate to the `ansible/ansible_roles` page in the UI. The new form automatically appears when I navigate to this page (without pressing anything), and the `Import from VCS...` button doesn't always work. 
In addition, when I enter a URL and select a smart-proxy, clicking on the `Examin` button suddenly closes the popup form.

nofaralfasi commented:
Unfortunately, this is still not working for me as accepted. 
I've messaged @Thorben-D the error I'm encountering. Please advise on how we should proceed from here.

nofaralfasi commented on app/lib/proxy_api/ansible.rb:
```diff
@@ -53,5 +53,62 @@ def playbooks(playbooks_names = [])
+      raise e unless e.is_a? RestClient::RequestFailed
+      case e.http_code
+      when 409
+        raise Foreman::Exception.new N_('A repo with the name %rName already exists.') % repo_info['name']
```
Shouldn't it be:
```suggestion
        raise Foreman::Exception.new N_('A repo with the name %rName already exists.') % repo_info['repo_info']['name']
```
And I would suggest adding the `&` operator: `repo_info['repo_info']&.[]('name')`.

Thorben-D replied:
Indeed. Also, good call on the safe navigation operator. Did not know it could be used that way.

nofaralfasi commented:
Here are some suggested improvements to the GUI:
1. Enhance the GUI by including an asterisk next to the mandatory fields.
2. When clicking on the `Examine` button with a non-existing Git link (in the correct format), the following issues occur:
   - No error message is displayed.
   - The application becomes unresponsive and requires a restart to function properly again.
3. Clicking on the `Confirm` button does not close the modal upon successful completion.
4. Enhance the form by adding an icon featuring a question mark, providing additional details about each field when hovered over.

nofaralfasi commented:
Thanks, @Thorben-D, for the improvements!

Other things I'm noticing:
- The tabs appear disabled even when they are enabled.
- The "SmartProxy required" message should be placed under the smart proxy selection.
- The blue border around the "Collision handling" switch is unnecessary. Additionally, I believe the text of the switch should remain "Update existing" and only the switch should be toggled on/off.

Renaming suggestions:
- Rename "Smart Proxy name" to "Choose Smart Proxy"
- Rename "Import from VCS" to "Clone from Git"
- Rename "Get Ansible roles from VCS" to "Clone Ansible roles from Git"
- "When toggled, Foreman will override already installed roles with the given name." to "When toggled, Foreman will overwrite existing roles with the same name."

nofaralfasi commented on app/helpers/foreman_ansible/ansible_roles_helper.rb:
```diff
@@ -31,6 +31,12 @@ def ansible_proxy_import(hash)
     end
 
+    def vcs_import
+      select_action_button("",
```
We should not display this button if the user does not have the required permissions.

Thorben-D replied:
Missed this as well. Will add later.

Thorben-D replied:
Actually, I am not sure how to handle this... Ideally, this would be handled on the React side.

nofaralfasi reviewed (changes requested):
Some actions, such as listing and deleting, are not directly related to the ability to import Ansible roles from Git. Perhaps I'm overthinking it, but I suggest we should decide whether to move these actions to the `AnsibleRolesController` or keep the structure as is.

Thorben-D commented on webpack/components/VcsCloneModalContent/components/BranchTagSelectionMenu.js:
```diff
@@ -0,0 +1,123 @@
+        eventKey={1}
+        title={<TabTitleText>{__('Branches')}</TabTitleText>}
+        isDisabled={!props.branchTagsEnabled}
+      >
```
This is sort-of happening. Since "main" is the default ref, the user would have to delete it from "manual input" to show the error message.

Thorben-D commented on webpack/components/VcsCloneModalContent/components/GitLinkInputComponent.js:
```diff
@@ -0,0 +1,174 @@
+    <React.Fragment>
+      <Form>
+        <FormGroup
+          label={__('Git-URL')}
```
Missed this, will do.

Thorben-D replied:
Do we have a fitting example for that link?

nofaralfasi replied:
Something like: `https://github.com/example/example.git`

Thorben-D commented on webpack/components/VcsCloneModalContent/components/SmartProxySelector.js:
```diff
@@ -0,0 +1,152 @@
+          onSelect={handleSmartProxySelect}
+          selections={props.smartProxySelection}
+          isOpen={isSmartProxyDropdownOpen}
+          placeholderText={__('Choose Smart Proxy')}
```
Missed this, will do.

Thorben-D replied:
Actually, I am not sure how to handle this... Ideally, this would be handled on the React side.

Thorben-D commented on app/views/ansible_roles/index.html.erb:
```diff
@@ -1,7 +1,14 @@
+
+<%= webpacked_plugins_js_for :foreman_ansible %>
+<%= webpacked_plugins_css_for :foreman_ansible %>
```
If i remove these two lines, the component cannot be mounted. Could you point me to a PR where this was deprecated?

MariaAga replied:
selecting both was a misclick, only meant to select the css one. it was deprecated here :https://github.com/theforeman/foreman/commit/89c610451c9ea22eb426826e405518e04d68972f

Thorben-D replied:
Yeah, that makes sense. I'll remove the one added by me and create a different PR for the rest.

nofaralfasi commented on webpack/components/VcsCloneModalContent/components/BranchTagSelectionMenu.js:
```diff
@@ -0,0 +1,123 @@
+            />
+          </FormGroup>
+        </Form>
+      </Tab>
```
I see the UI after the changes now, and I think the previous version looked better.
The default selected tab should be the rightmost one.

nofaralfasi commented:
Since this PR has been inactive for a while, we’re closing it to keep the repository organized. If you’d like to continue working on it, please feel free to reopen or create a new PR.
